File: tools/snake_trace.py
"""
tools/snake_trace.py

Python counterpart to tools/snake_trace.js.
Runs the SphericalSnakeEnv physics with the same deterministic LCG RNG and
outputs an identical JSON trace for diffing.

Usage:
    python tools/snake_trace.py --seed 42 --actions '[1,1,1,0,2,1]'

The LCG state machine is identical to snake_trace.js:
    state = (state * 1664525 + 1013904223) & 0xFFFFFFFF
    return state / 2**32
"""

import argparse
import json
import math
import sys
from pathlib import Path

# Allow importing from src/agent
sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import numpy as np
from agent.env import (
    SphericalSnakeEnv,
    NODE_ANGLE,
    NODE_QUEUE_SIZE,
    STARTING_DIRECTION,
    SNAKE_VELOCITY,
    TURN_RATE,
    INITIAL_SNAKE_LENGTH,
)


# ---------------------------------------------------------------------------
# Deterministic LCG — identical to snake_trace.js
# ---------------------------------------------------------------------------


class LCG:
    def __init__(self, seed: int) -> None:
        self.state = int(seed) & 0xFFFFFFFF

    def rand(self) -> float:
        self.state = (self.state * 1664525 + 1013904223) & 0xFFFFFFFF
        return self.state / 4294967296.0


# ---------------------------------------------------------------------------
# LCG-seeded subclass of SphericalSnakeEnv
# Overrides only _regenerate_pellet() to use the LCG.
# ---------------------------------------------------------------------------


class TracingEnv(SphericalSnakeEnv):
    def __init__(self, lcg: LCG) -> None:
        super().__init__()
        self._lcg = lcg

    def _regenerate_pellet(self) -> np.ndarray:
        """Mirrors JS regeneratePellet() using LCG instead of np_random."""
        theta = self._lcg.rand() * 2.0 * math.pi
        phi = self._lcg.rand() * math.pi
        sin_phi = math.sin(phi)
        return np.array(
            [math.cos(theta) * sin_phi, math.sin(theta) * sin_phi, math.cos(phi)],
            dtype=np.float64,
        )


def lcg_init(env: TracingEnv) -> None:
    """
    Replicate JS init() without going through gymnasium reset():
      1. Build pellet via LCG
      2. Build 8 snake nodes
    Mirrors the exact order in game/snake.js init().
    """
    env.direction = STARTING_DIRECTION
    env.score = 0
    env._terminated = False

    # JS calls regeneratePellet() BEFORE building snake nodes
    env.pellet = env._regenerate_pellet()

    env.snake = np.empty((0, 3), dtype=np.float64)
    env.pos_queues = np.empty((0, NODE_QUEUE_SIZE, 3), dtype=np.float64)
    for _ in range(INITIAL_SNAKE_LENGTH):
        env._add_snake_node()


# ---------------------------------------------------------------------------
# Snapshot helper — mirrors snake_trace.js snapshot()
# ---------------------------------------------------------------------------


def snapshot(env: TracingEnv, tick: int) -> dict:
    head = env.snake[0]
    # pos_queues[0] is now a (NODE_QUEUE_SIZE, 3) ndarray — all entries are real floats.
    # JS has null for unfilled slots (first 9 ticks); we output the pre-filled fallback
    # values instead.  compare_traces.py skips JS-side nulls, so these extra entries
    # are flagged as "MISSING in JS" for ticks 0-8 only.
    pq0 = [{"x": float(p[0]), "y": float(p[1]), "z": float(p[2])} for p in env.pos_queues[0]]
    return {
        "tick": tick,
        "direction": float(env.direction),
        "head": {"x": float(head[0]), "y": float(head[1]), "z": float(head[2])},
        "snake_length": len(env.snake),
        "pellet": {
            "x": float(env.pellet[0]),
            "y": float(env.pellet[1]),
            "z": float(env.pellet[2]),
        },
        "posQueue_0": pq0,
    }


# ---------------------------------------------------------------------------
# CLI harness
# ---------------------------------------------------------------------------


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--seed", type=int, default=42)
    parser.add_argument("--actions", type=str, default="[]")
    args = parser.parse_args()

    actions = json.loads(args.actions)
    lcg = LCG(args.seed)
    env = TracingEnv(lcg)
    lcg_init(env)

    trace = []
    for t, action in enumerate(actions):
        # --- Tick order mirrors JS update() while-loop exactly ---
        eaten, self_collision = env._check_collisions()

        if self_collision:
            snap = snapshot(env, t)
            snap["event"] = "terminated"
            trace.append(snap)
            break

        if action == 1:
            env.direction -= TURN_RATE
        elif action == 2:
            env.direction += TURN_RATE

        env._apply_snake_rotation()
        env._world_rotation()

        snap = snapshot(env, t)
        if eaten:
            snap["event"] = "pellet_eaten"
        trace.append(snap)

    print(json.dumps(trace, indent=2))


if __name__ == "__main__":
    main()

File: tools/snake_trace.js
/**
 * tools/snake_trace.js  (Option B)
 *
 * Loads docs/snake.js UNMODIFIED via vm.runInThisContext so its physics
 * functions and globals run in this process's global scope with zero copying.
 * Only changes from the browser environment:
 *   - Minimal DOM stubs satisfy snake.js's top-level querySelector/addEventListener calls
 *   - Math.random is replaced with a deterministic LCG before the script executes
 *   - window.requestAnimationFrame is a no-op (we drive ticks manually)
 *   - showEnd() still sets stopped=true; we read that flag instead of watching the DOM
 *
 * Usage:
 *   node tools/snake_trace.js --seed 42 --actions '[1,1,1,0,2,1]'
 *
 * Output: JSON array, one object per tick:
 *   { tick, direction, head, snake_length, pellet, posQueue_0 }
 */

'use strict';

const vm   = require('vm');
const fs   = require('fs');
const path = require('path');

// ---------------------------------------------------------------------------
// 1. Deterministic LCG — identical parameters to tools/snake_trace.py
// ---------------------------------------------------------------------------
var _lcgState = 0;

function _lcgSeed(s) { _lcgState = (s >>> 0); }

function _lcgRand() {
    _lcgState = ((Math.imul(_lcgState, 1664525) + 1013904223) >>> 0);
    return _lcgState / 4294967296;
}

// ---------------------------------------------------------------------------
// 2. Minimal DOM stubs — just enough for snake.js to load and run
// ---------------------------------------------------------------------------
function _makeElement() {
    var el = {
        style: {},
        innerHTML: '',
        classList: { add: function(){}, remove: function(){} },
        addEventListener: function(){},
    };
    return el;
}

var _canvasCtx = new Proxy({}, { get: function() { return function(){}; } });
var _canvasEl  = { getContext: function() { return _canvasCtx; }, width: 358, height: 360 };

global.document = {
    querySelector:        function()  { return _makeElement(); },
    getElementById:       function()  { return _makeElement(); },
    getElementsByTagName: function(t) {
        if (t === 'canvas') return [_canvasEl];
        return [_makeElement()];
    },
};

global.window = {
    addEventListener:     function() {},
    requestAnimationFrame:function() {},   // no-op — we drive ticks manually
    location: { reload: function() {} },
};

// ---------------------------------------------------------------------------
// 3. Seed LCG and replace Math.random BEFORE loading snake.js,
//    so init() → regeneratePellet() uses it from the very first call.
//    Seed is parsed from CLI args here, snake.js is loaded just below.
// ---------------------------------------------------------------------------
var _args    = process.argv.slice(2);
var _seed    = 42;
var _actions = [];

for (var _i = 0; _i < _args.length; _i++) {
    if (_args[_i] === '--seed')    _seed    = parseInt(_args[++_i], 10);
    if (_args[_i] === '--actions') _actions = JSON.parse(_args[++_i]);
}

_lcgSeed(_seed);
Math.random = _lcgRand;   // patch before load so init() → regeneratePellet() uses LCG

// ---------------------------------------------------------------------------
// 4. Load docs/snake.js in THIS global context.
//    vm.runInThisContext makes all its `var` globals (snake, pellet, direction,
//    stopped, snakeVelocity, …) and functions (rotateZ, rotateY,
//    applySnakeRotation, checkCollisions, …) accessible directly below.
//    init() at the bottom of snake.js runs here, building the initial state.
// ---------------------------------------------------------------------------
var _snakeJsPath = path.join(__dirname, '..', 'docs', 'snake.js');
vm.runInThisContext(fs.readFileSync(_snakeJsPath, 'utf8'), { filename: _snakeJsPath });

// After the above line, all of snake.js's globals are live in this scope:
//   snake, pellet, direction, stopped, snakeVelocity,
//   checkCollisions(), applySnakeRotation(), rotateZ(), rotateY(), etc.

// ---------------------------------------------------------------------------
// 5. Snapshot helper
// ---------------------------------------------------------------------------
function _snapshot(tick) {
    var pq0 = snake[0].posQueue.map(function(p) {
        return p ? { x: p.x, y: p.y, z: p.z } : null;
    });
    return {
        tick:         tick,
        direction:    direction,
        head:         { x: snake[0].x, y: snake[0].y, z: snake[0].z },
        snake_length: snake.length,
        pellet:       { x: pellet.x,   y: pellet.y,   z: pellet.z   },
        posQueue_0:   pq0,
    };
}

// ---------------------------------------------------------------------------
// 6. Drive ticks manually — mirrors the JS update() while-loop body exactly
// ---------------------------------------------------------------------------
(function main() {
    var trace = [];

    for (var t = 0; t < _actions.length; t++) {
        // step 1: collision check
        checkCollisions();

        if (stopped) {
            trace.push(Object.assign(_snapshot(t), { event: 'terminated' }));
            break;
        }

        // Was a pellet eaten this tick? checkCollisions() already mutated
        // snake_length and called regeneratePellet() if so — detect via
        // comparing length against previous tick.
        var prevLen = t === 0 ? 8 : trace[t - 1].snake_length;
        var eaten   = (snake.length > prevLen);

        // step 2: action → direction
        var action = _actions[t];
        if (action === 1) direction -= 0.08;
        if (action === 2) direction += 0.08;

        // step 3: move snake
        applySnakeRotation();

        // step 4: world rotation (rotateZ/rotateY with no pt arg use allPoints()
        //         internally, which includes grid points + pellet + snake + queues)
        rotateZ(-direction);
        rotateY(-snakeVelocity);
        rotateZ(direction);

        var snap = _snapshot(t);
        if (eaten) snap.event = 'pellet_eaten';
        trace.push(snap);
    }

    process.stdout.write(JSON.stringify(trace, null, 2) + '\n');
})();
